docs/source/scheduler.rst

```diff
@@ -0,0 +1,263 @@
+ * Applications within reserved capacity has rank less or equal to applications
+   above reserved capacity.
+ * Applications within reserved capacity has utilization < 1.
+
+The hierarchical merge allows to achieve the following design goals:
+
+ * Different tenants reserve capacity independently (Treadmill does not allow
+   for reservations to go beyond total capacity of the cell)
+ * Tenant hierarchy allows to offer unused capacity to the sibling first,
+   parents later. Consider someone running grid application, and creating dev
+   and prod sub allocations. If prod is underutilized, capacity will be offered
+   first to the test grid, then to other Treadmill users. Similarly, stopping
+   test grid will boost capacity for prod grid first (sibling).
+ * Favor production applications to test (qa, uat). Typical setup is to
+   configure all allocations with same base rank, but assign different
+   adjustment to allocations for different environments. Prod allocations are
+   configured with rank adjustment 3, uat with 2, qa with 1, dev - 0. As such,
+   global queue will be partitioned as following: prod applications (ordered by
+   utilization) running within reserved capacity, uat applications within
+   reserved capacity, … , dev applications within reserved capacity,
+   non-reserved. Such scheme ensures that in the event of catastrophic capacity
+   failure production application running within reserved capacity will be
+   evicted last.
+
+
+.. _section-topology-constraints:
+
+Server topology constraints
+---------------------------
+
+Servers in the cell are organized in a hierarchy that aims to reflect physical
+topology of the datacenter.
+
+Typical structure is server, rack, pod (rack is collection of servers, pod is
+collection of racks).
+
+Treadmill scheduler supports two placement strategies - stack and spread. The
+default strategy is spread (if you run two instances of web server for HA, you
+do not want them to end up on the same server). Placement strategy can be
+specified in the application manifest. Placement strategies can be different
+for different levels. For examples, one may want to spread application across
+servers but stack them on rack and pod level, for backend data locality.
+
+In addition to placement strategies, Treadmill allows to specify affinity limit
+constraints. These constraints will ensure that no more than N applications of
+same affinity are placed on the same server (rack, pod).
+
+Combining placement strategies with affinity limit constraints allows to
+express various HA/performance scenarios for application placement.
+
+
+.. _section-application-lifetime:
+
+Application lifetime constraints and reboot awareness
+-----------------------------------------------------
+
+Treadmill scheduler is aware of the expected lifetime (time till reboot) for
+servers in the cell. Typical reboot schedule can be once in three weeks,
+percentage of cell rebooting every day spread across 3 weeks.
```

Comment:
I'd reword.   The typical reboot schedule of a server might be once every three weeks, and within a cell, there is often a fixed/known percentage of the cell's servers rebooting on a given day.  (Is that true?) - Current reading is somewhat vague.

Reply:
reworded